File: SimulatedAnnealing.java
import java.io.*;
import java.util.*;

public class SimulatedAnnealing {

    public static double[][] readCoordinates(String filename) throws IOException {
        List<double[]> coordinates = new ArrayList<>();
        try (BufferedReader br = new BufferedReader(new FileReader(filename))) {
            String line;
            while ((line = br.readLine()) != null) {
                String[] parts = line.trim().split("\\s+");
                double x = Double.parseDouble(parts[0]);
                double y = Double.parseDouble(parts[1]);
                coordinates.add(new double[]{x, y});
            }
        }
        return coordinates.toArray(new double[0][]);
    }


    public static double[][] readDistances(String filename) throws IOException {
        List<double[]> distances = new ArrayList<>();
        try (BufferedReader br = new BufferedReader(new FileReader(filename))) {
            String line;
            while ((line = br.readLine()) != null) {
                String[] parts = line.trim().split("\\s+");
                double[] row = Arrays.stream(parts).mapToDouble(Double::parseDouble).toArray();
                distances.add(row);
            }
        }
        return distances.toArray(new double[0][]);
    }


    public static double calculateRouteLength(int[] route, double[][] distances) {
        double length = 0.0;
        for (int i = 0; i < route.length - 1; i++) {
            length += distances[route[i]][route[i + 1]];
        }
        length += distances[route[route.length - 1]][route[0]]; 
        return length;
    }


    public static int[] simulatedAnnealing(double[][] distances, double initialTemperature, double coolingRate, int iterations) {
        int n = distances.length;
        int[] currentRoute = new int[n];
        for (int i = 0; i < n; i++) {
            currentRoute[i] = i;
        }
        shuffleArray(currentRoute);

        int[] bestRoute = currentRoute.clone();
        double bestLength = calculateRouteLength(bestRoute, distances);

        double temperature = initialTemperature;

        while (temperature > 1) {
            for (int i = 0; i < iterations; i++) {
                int[] newRoute = currentRoute.clone();
                int index1 = (int) (Math.random() * n);
                int index2 = (int) (Math.random() * n);
                reverseSubarray(newRoute, Math.min(index1, index2), Math.max(index1, index2));

                double newLength = calculateRouteLength(newRoute, distances);
                double delta = newLength - calculateRouteLength(currentRoute, distances);

                if (delta < 0 || Math.random() < Math.exp(-delta / temperature)) {
                    currentRoute = newRoute;
                    if (newLength < bestLength) {
                        bestRoute = newRoute.clone();
                        bestLength = newLength;
                    }
                }
            }
            temperature *= coolingRate;
        }
        return bestRoute;
    }


    public static void shuffleArray(int[] array) {
        Random random = new Random();
        for (int i = array.length - 1; i > 0; i--) {
            int index = random.nextInt(i + 1);
            int temp = array[index];
            array[index] = array[i];
            array[i] = temp;
        }
    }

 
    public static void reverseSubarray(int[] array, int start, int end) {
        while (start < end) {
            int temp = array[start];
            array[start] = array[end];
            array[end] = temp;
            start++;
            end--;
        }
    }


    public static void main(String[] args) {
        try {
            double[][] distances = readDistances("Dist1.txt");
            double[][] coordinates = readCoordinates("coord1.txt");

            long startTime = System.nanoTime();

            int[] bestRoute = simulatedAnnealing(distances, 10000, 0.999, 12000);

            long endTime = System.nanoTime();

            double elapsedTimeInSeconds = (endTime - startTime) / 1_000_000_000.0;

            System.out.println("Best route found with Simulated Annealing: " + Arrays.toString(bestRoute)); 
            double bestLength = calculateRouteLength(bestRoute, distances); 
            System.out.println("Length of the best route (Simulated Annealing): " + bestLength);

            System.out.println("Execution time: " + elapsedTimeInSeconds + " seconds");

        } catch (IOException e) {
            System.err.println("Error reading files: " + e.getMessage());
        }
    }
}

File: Simulated Annealing.py
import numpy as np
import math
import random
import matplotlib.pyplot as plt

def read_coordinates(filename):
    with open(filename, 'r') as file:
        coordinates = []
        for line in file:
            x, y = map(float, line.strip().split())
            coordinates.append((x, y))
    return np.array(coordinates)

def read_distances(filename):
    with open(filename, 'r') as file:
        distances = [list(map(float, line.strip().split())) for line in file]
    return np.array(distances)

def calculate_route_length(route, distances):
    length = sum(distances[route[i], route[i + 1]] for i in range(len(route) - 1))
    length += distances[route[-1], route[0]]
    return length

def simulated_annealing(distances, initial_temperature=10000, cooling_rate=0.99, iterations=1100):
    n = len(distances)
    current_route = list(range(n))
    random.shuffle(current_route)
    best_route = current_route[:]
    best_length = calculate_route_length(best_route, distances)
    
    temperature = initial_temperature
    
    while temperature > 1:
        for _ in range(iterations):
            new_route = current_route[:]
            i, j = sorted(random.sample(range(n), 2))
            new_route[i:j+1] = reversed(new_route[i:j+1])
            
            new_length = calculate_route_length(new_route, distances)
            
            delta = new_length - calculate_route_length(current_route, distances)
            if delta < 0 or random.random() < math.exp(-delta / temperature):
                current_route = new_route[:]
                
                if new_length < best_length:
                    best_route = new_route[:]
                    best_length = new_length
        
        temperature *= cooling_rate
    
    return best_route, best_length



coordinates = read_coordinates("Coord1.txt")
distances = read_distances("Dist1.txt")


best_route, best_length = simulated_annealing(distances)
print("Best route found with Simulated Annealing:", best_route)
print("Length of the best route (Simulated Annealing):", best_length)
